File: src/java/magma/collect/list/List_.java
package magma.collect.list;

import magma.collect.stream.Stream;
import magma.option.Option;
import magma.option.Tuple;

import java.util.function.BiFunction;
import java.util.function.Consumer;
import java.util.function.Function;

public interface List_<T> {
    Stream<T> stream();

    List_<T> add(T element);

    List_<T> addAll(List_<T> others);

    Option<T> findFirst();

    int size();

    List_<T> subList(int start, int end);

    boolean equalsTo(List_<T> other);

    List_<T> sort(BiFunction<T, T, Integer> comparator);

    T get(int index);

    Option<Tuple<T, List_<T>>> popFirst();

    boolean isEmpty();

    List_<T> clear();

    Option<T> findLast();

    void forEach(Consumer<T> consumer);

    List_<T> mapLast(Function<T, T> mapper);
}

File: src/clang/magma/collect/list/List_.c
#include "List_.h"
magma.collect.stream.Stream<T> stream();
magma.collect.list.List_<T> add(T element);
magma.collect.list.List_<T> addAll(magma.collect.list.List_<T> others);
magma.option.Option<T> findFirst();
int size();
magma.collect.list.List_<T> subList(int start, int end);
int equalsTo(magma.collect.list.List_<T> other);
magma.collect.list.List_<T> sort(int(*comparator)(T, T));
T get(int index);
magma.option.Option<magma.option.Tuple<T, magma.collect.list.List_<T>>> popFirst();
int isEmpty();
magma.collect.list.List_<T> clear();
magma.option.Option<T> findLast();
magma.collect.list.void forEach(void(*consumer)(T));
magma.collect.list.List_<T> mapLast(T(*mapper)(T));

